internal/runtimehandlerhooks/high_performance_hooks_linux.go

```diff
@@ -204,17 +220,231 @@ func annotationValueDeprecationWarning(annotation string) string {
+}
+
+func addOrRemoveCpusetFromManagers(states []*desiredManagerCPUSetState, add bool) error {
+	allCPUs, err := fullCPUSet()
+	if err != nil {
+		return err
+	}
+
+	// Adding, we go top to bottom, when removing, bottom to top.
+	if !add {
+		slices.Reverse(states)
+	}
+	for _, state := range states {
+		// If we're updating a cpuset to be exclusive, we need to set cpuset.cpus first
+		if add {
+			// If the cgroup is shared among exclusive and not exclusive cgroups, then
+			// we should write the full cpuset. This allows the kernel to toggle the cpuset
+			// based on whether the cgroup uses partition mode "isolated", and ensures children
+			// cgroups that don't have load balance disabled can use the remaining cpus.
+			if state.shared {
+				if err := addOrRemoveCpusetFromManager(state.manager, allCPUs, add, cpusetCpus); err != nil {
+					return err
+				}
+			} else if err := addOrRemoveCpusetFromManager(state.manager, state.exclusiveCPUs, add, cpusetCpus); err != nil {
+				return err
+			}
+		}
+
+		// Unconditionally update cpuset.cpus.exclusive, as this file must contain any cpus that we intend to isolate.
+		if err := addOrRemoveCpusetFromManager(state.manager, state.exclusiveCPUs, add, cpusetCpusExclusive); err != nil {
+			return err
+		}
+
+		// If we're removing, we need to remove from cpuset.cpus after it's been removed from cpuset.cpus.exclusive
+		// Don't modify cpuset.cpus when shared. Not only is there no need, it doesn't even work because that cpuset has been
+		// written to the other children by the kernel.
+		if !add && !state.shared {
+			if err := addOrRemoveCpusetFromManager(state.manager, state.exclusiveCPUs, add, cpusetCpus); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func addOrRemoveCpusetFromManager(mgr cgroups.Manager, cpus cpuset.CPUSet, add bool, file string) error {
+	currentCpusStr, err := cgroups.ReadFile(mgr.Path(""), file)
+	if err != nil {
+		return err
+	}
+	currentCpus, err := cpuset.Parse(strings.TrimSpace(currentCpusStr))
+	if err != nil {
+		return err
+	}
+
+	var targetCpus cpuset.CPUSet
+	if add {
+		targetCpus = currentCpus.Union(cpus)
+	} else {
+		targetCpus = currentCpus.Difference(cpus)
+	}
+
+	// if we're writing to cpuset.cpus.exclusive, libcontainer manager doesn't have a field to manage it,
+	// so write to a file instead.
+	if file == cpusetCpusExclusive {
+		return cgroups.WriteFile(mgr.Path(""), file, targetCpus.String())
+	}
+	// otherwise, we should use the mgr directly, as it will go through systemd if necessary
+	return mgr.Set(&configs.Resources{
+		SkipDevices: true,
+		CpusetCpus:  targetCpus.String(),
+	})
+}
+
+var (
+	fullCPUSetOnce sync.Once
+	fullCPUSetErr  error
+	fullCPUSetVar  cpuset.CPUSet
+)
+
+// fullCPUSet returns the node's full CPUSet, which is used to populate the shared cgroups.
+// Do this just once. It's not particularly inefficient, but there's no need to recalculate.
```

Comment:
nonblocking, perhaps for a future cleanup: can't we just read `/sys/devices/system/cpu/online` ?

Reply:
this is much cleaner!